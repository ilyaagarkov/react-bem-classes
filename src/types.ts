import * as React from 'react';

export type AllowedModifierList = string[];
export interface PassedModifierHash {
  [name: string]: string;
}

export interface Config {
  blockName: string;
  props: {
    className?: React.HtmlHTMLAttributes<any>['className'];
    [key: string]: any;
  };
  allowedModifiers: string[];
}
