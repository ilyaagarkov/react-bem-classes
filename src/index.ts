export { bem } from './bem';
export { withBem } from './withBem';
export { Config, PassedModifierHash, AllowedModifierList } from './types';
