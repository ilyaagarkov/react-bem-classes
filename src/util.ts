import { PassedModifierHash } from './types';

interface CreateModifierOptions {
  baseClass: string;
  modifierName: string;
  modifierValue: string | boolean;
}

export const createModifier = ({ baseClass, modifierName, modifierValue }: CreateModifierOptions) => {
  if (!modifierValue || modifierValue === false) return '';

  let className = `${baseClass}--${modifierName}`;
  if (modifierValue && typeof modifierValue === 'string') className += `-${modifierValue}`;

  return className;
};

interface ModifiersFromObjOptions {
  baseClass: string;
  modifiers: PassedModifierHash;
}

export const modifiersFromObj = ({ baseClass, modifiers }: ModifiersFromObjOptions) =>
  Object.keys(modifiers)
    .map(modifierName =>
      createModifier({
        baseClass,
        modifierName,
        modifierValue: modifiers[modifierName],
      }),
    );
