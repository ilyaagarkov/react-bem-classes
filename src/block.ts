import { createModifier, modifiersFromObj } from './util';
import { Config, PassedModifierHash } from './types';

interface Options extends Config {
  passedModifiers: PassedModifierHash;
}

export const block = ({ blockName, props, passedModifiers, allowedModifiers }: Options) => {
  const classesSet = [] as string[];

  if (props.className) classesSet.push(props.className);

  classesSet.push(blockName);

  const modifiersFromProps = allowedModifiers
    .filter(modifierName => Boolean(props[modifierName]))
    .map(modifierName => createModifier({
      modifierName,
      modifierValue: props[modifierName],
      baseClass: blockName,
    }));

  const modifiersFromArguments = modifiersFromObj({
    baseClass: blockName,
    modifiers: passedModifiers,
  });

  classesSet.push(...modifiersFromProps, ...modifiersFromArguments);

  return classesSet.join(' ');
};
