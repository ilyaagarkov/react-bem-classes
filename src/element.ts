import { modifiersFromObj } from './util';
import { Config, PassedModifierHash } from './types';

interface Options {
  passedModifiers?: PassedModifierHash;
  blockName: Config['blockName'];
  elementName: string;
}

export const element = ({ blockName, passedModifiers = {}, elementName }: Options) => {
  const elementClass = `${blockName}__${elementName}`;
  const modifiersClasses = modifiersFromObj({
    baseClass: elementClass,
    modifiers: passedModifiers,
  });

  return [elementClass, ...modifiersClasses].join(' ');
};
