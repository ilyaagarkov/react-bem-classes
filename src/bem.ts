import { Config, PassedModifierHash } from './types';
import { block } from './block';
import { element } from './element';

export const bem = ({ blockName, props, allowedModifiers }: Config) => ({
  block: (passedModifiers?: PassedModifierHash) =>
    block({ blockName, props, passedModifiers, allowedModifiers }),

  element: (elementName: string, passedModifiers?: PassedModifierHash) =>
    element({ blockName, elementName, passedModifiers }),
});
