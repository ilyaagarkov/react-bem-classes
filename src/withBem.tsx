import * as React from 'react';

import { block as blok } from './block';
import { element } from './element';
import { AllowedModifierList, PassedModifierHash } from './types';

interface Options {
  block: string;
  modifiers: AllowedModifierList;
}

/* tslint:disable-next-line */
export const withBem = <P extends object>({ block, modifiers }: Options) => (Component: React.ComponentType<P>) => {
  return class ComponentWithBem extends React.Component<P> {
    block = (passedModifiers: PassedModifierHash) =>
      blok({ passedModifiers, blockName: block, props: this.props, allowedModifiers: modifiers })

    element = (elementName: string, passedModifiers: PassedModifierHash) =>
      element({ elementName, passedModifiers, blockName: block })

    render() {
      return <Component {...this.props} />;
    }
  };
};
