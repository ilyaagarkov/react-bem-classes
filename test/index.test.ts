import * as BEM from '../src/index';

describe('BEM', () => {
  it('provides correct API', () => {
    expect(BEM).toMatchSnapshot();
  });
});
