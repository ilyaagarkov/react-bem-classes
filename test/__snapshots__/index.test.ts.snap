// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BEM provides correct API 1`] = `
Object {
  "bem": [Function],
  "withBem": [Function],
}
`;
